### DISCUSSION TITLE
TW-elements carousel doesn't slide as in example

### DISCUSSION THREAD

**Discussion by jayrdi** [mdbootstrap/TW-Elements#2183] [Support from community]
I have a carousel on my site taken from [this example](https://tw-elements.com/docs/standard/components/carousel/#slides_only) but, whereas in the example the images slide across from right to left, on my site they simply change with no animation.

Does anyone know what is causing this, and how I might implement it so that it has that "slide" effect?

Thanks

**Comment by juujisai** [comment]
Hi! I've checked on fresh vite install and here in snippets and it slides with an animation for me: https://tw-elements.com/snippets/tailwind/b-cylwik/5817708

Does tailwind create those classes for you? (you can check inspect tools for that). If they are added the animation should work properly.
![image](https://github.com/mdbootstrap/TW-Elements/assets/52455076/695741ce-b34f-40a3-999b-d6314fd4dd9c)

If everything with the classes seems okay could you please show us some code you are using? Or is it the same exact code that is included bellow the `Slides Only` example?

Lastly, what stack are you using for your app?

**Reply by jayrdi** [reply]
Thanks @juujisai, 

Yes, I can see the classes correctly inspecting in the browser. Perhaps I'm missing something elsewhere.

I've built this with Statamic CMS and running on Hetzner server via ploi.io

Here is my code:

```
<div class="relative"
  data-te-carousel-init
  data-te-ride="carousel">
  <!--Carousel items-->
    <div class="relative w-full overflow-hidden after:clear-both after:block after:content-['']">
        <!--First item-->
        <div class="relative float-left -mr-[100%] w-full transition-transform duration-[400ms] ease-in-out motion-reduce:transition-none -z-10"
        data-te-carousel-item
        data-te-carousel-active>
            <img src="{{ header_image:0 }}" class="block w-full" alt="{{ header_image:0:alt }}"/>
        </div>
        <!--Second item-->
        <div class="relative float-left -mr-[100%] hidden w-full transition-transform duration-[400ms] ease-in-out motion-reduce:transition-none -z-10"
        data-te-carousel-item>
            <img src="{{ header_image:1 }}" class="block w-full" alt="{{ header_image:1:alt }}" />
        </div>
        <!--Third item-->
        <div class="relative float-left -mr-[100%] hidden w-full transition-transform duration-[400ms] ease-in-out motion-reduce:transition-none -z-10"
        data-te-carousel-item>
            <img src="{{ header_image:2 }}" class="block w-full" alt="{{ header_image:2:alt }}" />
        </div>
        <div class="flex flex-col justify-center items-center mt-44">
            <h4 class="mt-5 text-center text-3xl text-white">The</h4>
            <h1 class="text-center text-7xl text-white font-black drop-shadow-lg leading-normal">
                John Walton Muscular Dystrophy<br>Research Centre
            </h1>
            <p class="mt-6 mb-12 text-center max-w-xl font-semibold text-2xl text-white">{{ subtitle }}</p>
            <a href="#who-are-we" class="text-center transition hover:scale-[1.05]">
                <i class="fa-solid fa-circle-chevron-down text-white text-4xl"></i>
                <p class="text-white font-semibold mb-6">Find out more</p>
            </a>
        </div>
    </div>
</div>
```
Thanks for your help!

**Reply by juujisai** [reply]
I've checked your code on a fresh vanilla js install and the animations are working for me so the classes are not the problem here. Other than the animations the component is working correctly? If it didn't work at all then it would mean that there were problems with initialization.

There is a chance that you have disabled animations on your system but you said that on our docs the items are animating correctly right?. Are any other components animating? To make sure, try removing `motion-reduce:transition-none` class from elements with `data-te-carousel-item` attribute or check whether or try to check the system settings if animations are not turned off.

**Reply by jayrdi** [reply]
Yeah, my code with the photos from the tw-elements example works correctly in the "snippets" application when I run it... so it can't be my browser, it must be something to do with Statamic? As it only seems to "break" when it's running within my Statamic build...

Everything else looks correct, and the images do change, they just switch without the scrolling animation, which is strange

**Reply by juujisai** [reply]
Did you check if other components are having similar problems with animations? Like dropdowns, modals, steppers or is it just the carousel?

**Reply by jayrdi** [reply]
Just the carousel. I have a stepper and flush accordion also on the site which animate correctly..

**Reply by juujisai** [reply]
I've tried to play a bit with this today and found out that similar issue is happenning when I try to initialize the Carousel component via JS and do not provide the `data-te-carousel-init` attribute

Is this how it looks like in your app?
https://tw-elements.com/snippets/tailwind/b-cylwik/5822583#js-tab-view

Since you do have the attribute added then the issue is not the same for you but what I wanted to suggest is to try to initialize the component via JS and see if that could help here. First we would have to check whether an old instance exist and dispose it since I think you are using the `umd` format of the package:

```
const carouselEl = document.getElementById("carousel")

const oldInstance = te.Carousel.getInstance(carouselEl)
oldInstance && oldInstance.dispose();

new te.Carousel(carouselEl)

```

If you are using the ES package then you can skip the `oldInstance` check and not use `initTE` method there.

I hope that helps.

**Reply by jayrdi** [reply]
Thank you so much for your help with this! You're right, that's how it appears in my app, with no "sliding" effect.

I am initialising via `initTE` in `site.js`:

```
import {
  Carousel,
  Collapse,
  Dropdown,
  initTE,
} from "tw-elements";

initTE({ Carousel, Collapse, Dropdown }, { allowReinits: true });
```
I have `{{ vite src="resources/js/site.js|resources/css/site.css" }}` in my base template (although I don't think there's anything in there that affects this).

Under `tailwind.config.js` I have:

```
module.exports = {
  content: [
    ...
    "./src/**/*.{html,js}",
    "./node_modules/tw-elements/dist/js/**/*.js",
  ],
plugins: [
    require('@tailwindcss/typography'),
    require("tw-elements/dist/plugin.cjs"),
  ],
  darkMode: "class"
```

Does that all look correct? Are you suggesting I get rid of the "Carousel" bit from the existing `initTE` in `site.js` and instead put your block of JS underneath?

Thanks!

**Reply by juujisai** [reply]
Did initializing via JS work for you? Everything looks good but since you are using ES package then we don't have to dispose the instance first. Just remove `Carousel` from the `initTE`

```
import {
  Carousel,
  Collapse,
  Dropdown,
  initTE,
} from "tw-elements";

initTE({ Collapse, Dropdown }, { allowReinits: true });

const carouselEl = document.getElementById("carousel-1");
new Carousel(carouselEl)
```

Just make sure that the component is initialized after route changes

**Reply by jayrdi** [reply]
Ah, so actually animations aren't working correctly in other elements. It does work properly with the "flush accordion", which is why I thought it was okay, but I just tested with the "drop in" (https://tw-elements.com/docs/standard/content-styles/animations-extended/) animation and that won't work.

I have added the class to "safelist", here is my `tailwind.config.js` (removed irrelevant lines):

```
const defaultTheme = require('tailwindcss/defaultTheme')
const { withAnimations } = require('animated-tailwindcss')

module.exports = withAnimations ({
  content: [
    ...
    "./src/**/*.{html,js}",
    "./node_modules/tw-elements/dist/js/**/*.js",
  ],
  theme: {
    extend: {
      ...
    },
    safelist: 
      ['animate-[drop-in_0.5s]'],
  },
  plugins: [
    require("tw-elements/dist/plugin.cjs"),
  ],
  darkMode: "class"
})
```
I'm using Vite if that makes any difference? `vite.config.js`:

```
import { defineConfig } from 'vite';
import laravel from 'laravel-vite-plugin';

export default defineConfig({
    plugins: [
        laravel({
            input: [
                'resources/css/site.css',
                'resources/js/site.js',
            ],
            refresh: true,
        }),
    ],
});
```
My JS referenced earlier is in `site.js` so that all seems to be correct. I mean, the functionality for my TW-elements stuff works, it seems to just be the animations that are failing.

I don't have a `<script>` link in my base HTML template, as I figure I don't need one because Vite is initialising the JS file.

Sorry, I'm not sure how to make sure that the component is initialising after route changes...

Thanks!

**Reply by juujisai** [reply]
Adding animation classes to safelist is necessary because we are using here dynamic arbitrary classes that tailwind on itself would ignore - thats why it wasn't working. 
https://tw-elements.com/docs/standard/content-styles/animations/

As of the initialising after route changes, that may be an issue only for SPA so if the component works fine then there is no need to do anything with that.

You can try to initialize the component via JS like in my previous comments inside the `site.js` file. Did you try that? Was there any improvement or no differences at all?

**Reply by jayrdi** [reply]
No effect initialising the component via JS, unfortunately, other than losing my "back to top" button that appears on a page when you scroll down. That now only appears on the homepage and no other pages. My `site.js` currently:

```
// Tailwind Elements
import {
  Carousel,
  Collapse,
  Animate,
  Ripple,
  Dropdown,
  initTE,
} from "tw-elements";

initTE({ Collapse, Animate, Ripple, Dropdown }, { allowReinits: true });

// Get the button
const mybutton = document.getElementById("btn-back-to-top");
const carouselEl = document.getElementById("carousel-1");
new Carousel(carouselEl)

// When the user scrolls down 20px from the top of the document, show the button

const scrollFunction = () => {
  if (
    document.body.scrollTop > 20 ||
    document.documentElement.scrollTop > 20
  ) {
    mybutton.classList.remove("hidden");
  } else {
    mybutton.classList.add("hidden");
  }
};
const backToTop = () => {
  window.scrollTo({ top: 0, behavior: "smooth" });
};

// When the user clicks on the button, scroll to the top of the document
mybutton.addEventListener("click", backToTop);

window.addEventListener("scroll", scrollFunction);
```

As something of interest though, the "drop-in" box animation I dropped into the top of one of the pages as a test is still not working... I have it on the page as:

```
<div
    data-te-animation-init
    data-te-animation-start="onHover"
    data-te-animation-reset="true"
    data-te-animation="[drop-in_0.5s]"
    class="m-[20px] h-[130px] w-[130px] origin-[top_center] bg-[#fbfbfb] pt-10 text-center shadow-[0_2px_10px_0_rgba(0,0,0,0.14)] dark:bg-neutral-600">
    drop in
</div>
```
 and in `tailwind.config.js` as:

```
const defaultTheme = require('tailwindcss/defaultTheme')

module.exports = {
  content: [
    './resources/**/*.antlers.html',
    './resources/**/*.antlers.php',
    './resources/**/*.blade.php',
    './resources/**/*.vue',
    './content/**/*.md',
    "./src/**/*.{html,js}",
    "./node_modules/tw-elements/dist/js/**/*.js",
  ],
  theme: {
    extend: {
      fontFamily: {
        'sans': ['Montserrat', ...defaultTheme.fontFamily.sans],
      },
      colors: {
        'jwmdrc-blue' : '#193d6b',
        'jwmdrc-red' : '#cf122e',
        'jwmdrc-light-blue' : '#7cadd3',
      },
      typography: {
        DEFAULT: {
          css: {
            a: {
              'text-decoration': 'none',
              'border-bottom-width': '2px',
              'border-color': 'rgb(207 18 46)',
              'font-weight': '700',
              '&:hover': {
                'border-bottom-width': '4px',
                'background-color': 'rgba(207, 18, 46, 0.25)',
              },
              '&:focus': {
                'border-bottom-width': '4px',
                'border-color': 'rgb(0 0 0)',
                'background-color': 'rgb(250 204 21)',
              },
            },
          },
        },
      },
    },
    safelist: [
      'animate-[drop-in_0.5s]',
    ]
  },
  plugins: [
    require('@tailwindcss/typography'),
    require("tw-elements/dist/plugin.cjs"),
  ],
  darkMode: "class"
}
```

The `safelist` is implemented correctly? I just wonder if it's all related, as that's two animations that don't work correctly.

**Reply by juujisai** [reply]
The safelist is implemented correctly, it works fine for me in my fresh app. 

Maybe try checking again whether the classes are receiveing proper styles in inspect tools. The carousel animation is based on the `transition-transform` class and animations are based on `animation` classes - are the styles for those applied / created ?

You can also try linking our css (but make sure it doesn't overwrite your own styles) and see if anything changes
```
<link
  rel="stylesheet"
  href="https://cdn.jsdelivr.net/npm/tw-elements/dist/css/tw-elements.min.css" />
```
or via js
```
import "tw-elements/dist/css/tw-elements.min.css";
```

You can also try to check whether your those issues occur in fresh vanilla app.

Other than that, I'm not sure what may cause those issues and how to help. Most likely it's some kind of config or the environment causing them I think.
